### PR TITLE
Add support unit tests for ConcurrentHashSet, #1117

### LINKED ISSUE #1117: Add Support unit tests for ConcurrentHashSet
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Task description

From #920: Apache Harmony should have tests for this. Note that not all `ISet<T>` members are implemented. If they were, we could potentially move this to J2N. Note that 1 test was added in [#938](https://github.com/apache/lucenenet/pull/938/files#r1597522965), but we should be more thorough than this.

### PR DESCRIPTION
**Pull request by paulirwin**
- [X] You've read the [Contributor Guide](https://github.com/apache/lucenenet/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://www.apache.org/foundation/policies/conduct.html).
- [X] You've included unit or integration tests for your change, where applicable.
- [X] You've included inline docs for your change, where applicable.
- [X] There's an open issue for the PR that you are making. If you'd like to propose a change, please [open an issue](https://github.com/apache/lucenenet/issues/new/choose) to discuss the change or find an existing issue.

Add support unit tests for ConcurrentHashSet

Fixes #1117

## Description

This ports tests from Harmony for `Collections.synchronizedSet` and `ConcurrentHashMap` and adapts them to our `ConcurrentHashSet`.

This also adds nullable reference type checking to ConcurrentHashSet, which exposed some nullability issues. Given that our current use of this type is for values that should never be null (or would otherwise throw an NRE upon use as null is not checked-for), this adds a generic type constraint of `notnull` to ensure that it cannot be used with nullable reference types. Work would need to be done (under proper unit test coverage) to make this null-friendly if that is needed in the future.

### CODE REVIEW

**Review by NightOwl888 — CHANGES_REQUESTED** [review]
Thanks for the PR.

I have been considering phasing out the `SystemTypesHelpers` class (or at least most of the extension methods in it), so ideally any new tests we add would not have any dependencies on those extension methods (`.add()`, `.contains()`, `.size()`, etc.).

If we do keep them, they will definitely need a review. Some of the methods perform poorly or have better alternatives (`.toString()`, `removeAll()`, `retainAll()`, and `ToByteArray()` to name a few). The `.append()` overloads are generally there to allow culture invariant numbers to be appended, since .NET doesn't have that built-in. However, I am adding optional parameters to allow passing `format` and `IFormatProvider` on `OpenStringBuilder` in J2N. So, we should probably hold off on the review/removal of `SystemTypesHelpers` until after J2N has a build with `OpenStringBuilder` where those can be replaced directly by swapping in `OpenStringBuilder` for `StringBuilder`.

File: src/Lucene.Net/Support/ConcurrentHashSet.cs [hunk 1]
```diff
@@ -42,7 +43,8 @@ namespace Lucene.Net.Support
     /// concurrently from multiple threads.
     /// </remarks>
     [DebuggerDisplay("Count = {Count}")]
-    internal class ConcurrentHashSet<T> : ISet<T>, IReadOnlyCollection<T>, ICollection<T>
+    internal class ConcurrentHashSet<T> : ISet<T>, IReadOnlyCollection<T>
+        where T : notnull
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/Lucene.Net/Support/ConcurrentHashSet.cs:47]: It's more than just avoiding adding tests; the type as-is has nullability issues if T is allowed to be null. (You can see these if you pull this branch and remove this constraint.) These would have to be resolved (as well as adding test coverage for that). I didn't think that was needed in this case, given that our only usage of the type is where T is not nullable.

[the remark this answers is not in the document]
Comment (NightOwl888) [reply src/Lucene.Net/Support/ConcurrentHashSet.cs:47]: Looks like all of the warnings are because `IComparer<T>.GetHashCode()` doesn't allow `null`. This can be patched by creating a wrapper method. In other collections, `null` always corresponds to the hash code 0.

```c#
[MethodImpl(MethodImplOptions.AggressiveInlining)]
private int GetItemHashCode(T item) => item is not null ? _comparer.GetHashCode(item) : 0;
```

Do note that the nullability constraint doesn't match any of the BCL implementations, which return 0 for `null` rather than throwing. But doing it this way would future-proof it for any 3rd party `IEqualityComparer<T>` implementations, as well.

[the remark this answers is not in the document]
Comment (paulirwin) [reply src/Lucene.Net/Support/ConcurrentHashSet.cs:47]: I had tried that inline (instead of as a wrapper method) and thought I was running into some other issues, but that might have been because of other WIP. That fixed it, thanks.

File: src/Lucene.Net.Tests/Support/TestConcurrentHashSet.cs [hunk 2]
```diff
@@ -47,5 +171,278 @@ public void TestExceptWith()
...
+            // mySet.add(null);
+            // assertTrue("Trying to use nulls in list failed", mySet.contains(null));
+
+            smallSet = new HashSet<object>();
+            for (int i = 0; i < 100; i++)
+            {
+                smallSet.add(objArray[i]);
+            }
+            // LUCENENET TODO: could port this class and all of the classes it uses
+            // new Support_SetTest(new ConcurrentHashSet<object>(smallSet))
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/Lucene.Net.Tests/Support/TestConcurrentHashSet.cs:439]: That code appears to be GPL licensed, based on the license in the test project (and that there is no other license listed in this file). Would that be an issue? How would I do the license header in that case?

But we could always leave as a future TODO to port those tests. This test class gives very good test coverage already without them.

[the remark this answers is not in the document]
Comment (NightOwl888) [reply src/Lucene.Net.Tests/Support/TestConcurrentHashSet.cs:460]: The `Support_` classes were ported from Apache Harmony, so no issue with those:

https://github.com/apache/harmony/blob/trunk/classlib/support/src/test/java/tests/support/Support_SetTest.java
https://github.com/apache/harmony/blob/trunk/classlib/support/src/test/java/tests/support/Support_CollectionTest.java

The GPL license was put there because *some* of the test code is directly from the JDK and it is the most restrictive license. But we don't ship the test projects. J2N is made up from Apache 2.0 and MIT licenses and should all have headers, but I haven't gone through all of the test files to add license headers.

[the remark this answers is not in the document]
Comment (paulirwin) [reply src/Lucene.Net.Tests/Support/TestConcurrentHashSet.cs]: This is complete.

**Review by NightOwl888 — CHANGES_REQUESTED** [review]

File: src/Lucene.Net/Support/ConcurrentHashSet.cs [hunk 3]
```diff
@@ -42,7 +44,7 @@ namespace Lucene.Net.Support
     /// concurrently from multiple threads.
     /// </remarks>
     [DebuggerDisplay("Count = {Count}")]
-    internal class ConcurrentHashSet<T> : ISet<T>, IReadOnlyCollection<T>, ICollection<T>
+    internal class ConcurrentHashSet<T> : ISet<T>, IReadOnlyCollection<T>
```
[the remark this answers is not in the document]
Comment (paulirwin) [reply src/Lucene.Net/Support/ConcurrentHashSet.cs]: We already had FEATURE_READONLYSET defined for .NET 5+, so went ahead and added `IReadOnlySet<T>`. Added a `// ReSharper disable RedundantExtendsListEntry` to prevent an analysis warning about the redundant interface.

**Review by NightOwl888 — APPROVED** [review]
